docs/docs/development/events.md

```diff
@@ -0,0 +1,178 @@
+
+### Subscription and Listener
+
+To subscribe to any events, you will first need to inform the event manager that you wish to add a new listener.
+This is done by calling the `ZMK_LISTENER` macro:
+
+```c
+ZMK_LISTENER(combo, behavior_combo_listener);
+```
+
+This macro takes two parameters:
+
+1.  (`combo` in the example) This gives a name to the listener, for the event manager to refer back to it.
+2.  (`behavior_combo_listener` in the example) This is a [callback](<https://en.wikipedia.org/wiki/Callback_(computer_programming)>) that will be called whenever **any** event that the listener subscribes to occurs (if it is not handled by another listener with a higher priority). By convention, the callback should have the suffix `_listener`.
+
+Once you have a listener set up, you can subscribe to individual events by calling the `ZMK_SUBSCRIPTION` macro:
+
+```c
+ZMK_SUBSCRIPTION(combo, zmk_keycode_state_changed);
+```
+
+The first parameter is the name of the listener created with `ZMK_LISTENER`, while the second is the name of the _struct_ that defines the event's data, which was declared in the corresponding header file. By convention the header file for an event will be named `specific_thing_happened`, with the struct named `zmk_specific_thing_happened`.
+
+Of course, you will also need to import the corresponding event header at the top of your file.
+
+### Listener Callback
+
+The listener will be passed a raised `zmk_event_t` pointer (as described previously) as an argument, and should have `int` as its return type. By convention, the `zmk_event_t` is called
+
+The listener should return one of three values (which are of type `int`) back to the event manager:
+
+- `ZMK_EV_EVENT_BUBBLE`: Keep propagating the event `struct` to the next listener.
+- `ZMK_EV_EVENT_HANDLED`: Stop propagating the event `struct` to the next listener. The event manager still owns the `struct`'s memory, so it will be `free`d automatically. Do **not** free the memory in this function.
+- `ZMK_EV_EVENT_CAPTURED`: Stop propagating the event `struct` to the next listener. The event `struct`'s memory is now owned by your code, so the event manager will not free the event `struct` memory. Make sure your code will release or free the event at some point in the future. (Use the `ZMK_EVENT_*` macros described [below](#raising-events).)
+
+If an error occurs during the listener call, it should of course return the error.
+
+As mentioned previously, the same callback will be called when any event that is subscribed to occurs. To obtain the underlying event from the generic event passed to the listener, the previously described `as_zmk_specific_thing_happened` function should be used:
+
+```c
+int behavior_hold_tap_listener(const zmk_event_t *eh) {
+    if (as_zmk_position_state_changed(eh) != NULL) {
+        return position_state_changed_listener(eh);
+    } else if (as_zmk_keycode_state_changed(eh) != NULL) {
+        return keycode_state_changed_listener(eh);
```

Comment:
```suggestion
    if (as_zmk_position_state_changed(eh) != NULL) {
        // it is a position_state_changed event, handle it with my_position_state_handler
        return my_position_state_handler(eh);
    } else if (as_zmk_keycode_state_changed(eh) != NULL) {
        // it is a keycode_state_changed event, handle it with my_keycode_state_handler
        return my_keycode_state_handler(eh);
```

This snippet didn't make it clear to me that `*_listener` functions are custom and not provided by ZMK.